src/bottom_up_iterator.jl

```diff
@@ -12,235 +12,212 @@ Concrete implementations of this iterator should implement the following methods
 abstract type BottomUpIterator <: ProgramIterator end
 
 Base.@doc """
-    @programiterator BasicIterator(problem::Problem{Vector{IOExample}}) <: BottomUpIterator
+    @programiterator DepthIterator(problem::Problem{Vector{IOExample}}) <: BottomUpIterator
 
 A basic implementation of the bottom-up iterator. It will enumerate all programs in increasing order based on their depth.
-""" BasicIterator
-@programiterator BasicIterator(problem::Problem{Vector{IOExample}}) <: BottomUpIterator
+""" DepthIterator
+@programiterator DepthIterator(problem::Problem{Vector{IOExample}}) <: BottomUpIterator
 
-"""
-	struct BottomUpState
-
-Holds the state of the bottom-up iterator. This includes the priority bank, the hashes of the outputs of the programs, and the current programs that are being enumerated.
-"""
-struct BottomUpState
-    priority_bank::Dict{Symbol, Dict{RuleNode,Int64}}
-    hashes::Set{UInt}
-    current_programs::Queue{RuleNode}
+mutable struct BottomUpState
+    bank::Any
+    data::Any
```

Comment:
non-urgent thing to think about: should we leave `bank` and `data` as `Any`, or should we introduce some subtyping? I assume the bank will always be either a `Dict` so that programs are ordered over their type, or some kind of priority as in Brute. If we figure out more precise type, the code might be faster.

Reply:
I have decided to introduce two new abstract types: `BUDepthBank` and `BUDepthData` that the concrete iterators will need to extend. Please let me know what you think about this.